Customizing/global/plugins/Services/Repository/RepositoryObject/ReportTrDemandAdv/classes/class.ilObjReportTrDemandAdv.php

```diff
@@ -75,87 +81,218 @@ protected function buildQuery($query) {
+		$tf = new \CaT\Filter\TypeFactory();
+		$f = new \CaT\Filter\FilterFactory($pf, $tf);
+		$txt = function($id) { return $this->plugin->txt($id); };
+
+		return
+		$f->sequence
+		(
+			$f->sequence
+			(
+
+				/* BEGIN BLOCK - PERIOD */
+				$f->dateperiod
+				(
+					$txt("period")
+					, ""
+				)->map
+					(
+						function($start,$end) use ($f)
+						{
+							$pc = $f->dateperiod_overlaps_predicate
+							(
+								"crs.begin_date"
+								,"crs.begin_date"
+							);
+							return array ("date_period_predicate" => $pc($start,$end)
+										 ,"start" => $start
+										 ,"end" => $end);
+						},
+						$tf->dict
+						(
+							array
+							(
+								"date_period_predicate" => $tf->cls("CaT\\Filter\\Predicates\\Predicate")
+								,"start" => $tf->cls("DateTime")
+								,"end" => $tf->cls("DateTime")
 							)
-			->static_condition('crs.begin_date >= '.$this->gIldb->quote(date('Y-m-d'),'text'))
-			->static_condition("(crs.is_cancelled != 'Ja' OR crs.is_cancelled IS NULL)")
-			->static_condition('crs.hist_historic = 0')
-			->static_condition("crs.is_template = 'Nein'")
-			->static_condition("crs.begin_date != '0000-00-00'")
-			->static_condition($this->gIldb->in('crs.type',array('Webinar','Präsenztraining','Virtuelles Training'),false,'text'))
-			->static_condition($local_condition)
-			->static_condition('tpl.hist_historic = 0')
-			->static_condition('tpl.is_template = '.$this->gIldb->quote('Ja','text'))
-			->action($this->filter_action)
-			->compile();
-		return $filter;
+						)
+					),
+					/* END BLOCK - PERIOD */
+
+
+				/* BEGIN BLOCK - FILTER TOPICS */
+				$f->multiselectsearch
+				(
+					$txt("filter_topics")
+					, ""
+					, $this->getTopics()
+				),
+				/* END BLOCK - FILTER TOPICS */
+
+
+				/* BEGIN BLOCK - TRAINING TYPE */
+				$f->multiselectsearch
+				(
+					$txt("training_type")
+					, ""
+					, array('Webinar' => 'Webinar','Präsenztraining' => 'Präsenztraining','Virtuelles Training' => 'Virtuelles Training')
+				),
+				/* END BLOCK - TRAINING TYPE */
+
+
+				/* BEGIN BLOCK - STATUS */
+				$f->multiselectsearch
+				(
+					$txt("status")
+					, ""
+					, array('min_participants > bookings' => $txt('cancel_danger'), 'min_participants <= bookings' => $txt('no_cancel_danger'))
+				),
+				/* END BLOCK - STATUS */
+
+				/* BEGIN BLOCK - WAITING LIST */
+				$f->multiselectsearch
+				(
+					$txt("waiting_list")
+					, ""
+					, array("crs.waitinglist_active = 'Ja'" => $txt('waiting_list'), "crs.waitinglist_active = 'Nein'" => $txt('no_waiting_list'))
+				),
+				/* END BLOCK - WAITING LIST */
+
+				/* BEGIN BLOCK - BOOKING OVER */
+				$f->multiselectsearch
+				(
+					$txt("booking_over")
+					, ""
+					, array($db->quote(date('Y-m-d'),'text')." > booking_dl " => $txt('book_dl_over'),
+							$db->quote(date('Y-m-d'),'text')." <= booking_dl " => $txt('book_dl_not_over'))
+				)
+				/* END BLOCK - BOOKING OVER */
+
+			)->map
```

Comment:
kannst du bitte dieses mapping auf den höchste ebene schieben? dann sparst du dir einen index in den selections